torchdynamo/guards.py

```diff
@@ -242,11 +245,21 @@ def NN_MODULE(self, guard: Guard):
         self.ID_MATCH(guard)
         ref = self.arg_ref(guard)
         val = self.get(guard.name)
-        assert istype(val.training, bool)
-        if val.training:
-            self.code.append(f"{ref}.training")
+
+        def setup_guard():
+            assert istype(val.training, bool)
+            self.code.append(f"___training_state({ref}, {val.training})")
+
+        if hasattr(val, "_awaiting_init"):
```

Comment:
Would changing this to `hasattr(val, "training")` allow us to remove the `_awaiting_init` patching?

Patching `__new__` results in some noticeable slowdowns.